misk/src/main/kotlin/misk/backoff/Retries.kt

```diff
@@ -3,9 +3,17 @@ package misk.backoff
 /**
  * Retries the provided function up to a certain number of times, applying the given backoff
  * between each retry. The retry function is provided with current retry count, in case this is
- * relevant
+ * relevant.
+ *
+ * retryException returns true if the caught exception should be retried, subject to the limit.
+ * By default all exceptions are retried.
```

Comment:
I'm not convinced we want to retry exceptions by default. While it is a nice programming model (abstracts away failures), it has a real impact on the performance of the production system (especially when things aren't going well).

I'd like to spend some time thinking through other options and discussing the tradeoffs.

An alternative design is to never retry transactions and kick the exception up to the caller to retry.

* I think the caller needs to handle failures anyways, since many things can fail, not just database queries. Ideally we just make actions idempotent and they are safe to retry.
* Gives us natural jitter, since retries would probably come over the network
* It's easier to reason about retries in the entire system, since we haven't layered nested retries deep down in the stack.

This is just an alternative idea, but we should spend some time thinking through the trade offs.

Reply:
Definitely worth discussion. Lets talk on Wednesday’s sync.